data/tutorials/language/3ds_02_map.md

```diff
@@ -8,151 +8,209 @@ category: "Data Structures"
+
+### Checking if a key is contained in a Map
 
-```ocaml
-# let data = Stock.add Apple 10 data;;
-val data : int Stock.t = <abstr>
+To check if a key is contained in a Map we can use the `mem` function:
+
+```ocaml=
+let is_paguzar_in_the_map: bool =
+  StringMap.mem "paguzar" lucky_numbers
+;;
 ```
 
-We have now created a new map—again called `data`, thus masking the previous
-one—by adding `Apple` and its quantity `10` to our previous empty map.
-There is a fairly important point to make here. Once we have added the
-value `10` we have fixed the types of mappings that we can do.
-This means our mapping in our module `Stock` is from fruit _to int_.
-If we want a mapping from fruits to strings, we will have to create a different mapping.
+### Finding entries in a Map
 
-Let's add in some additional data just for kicks.
+To find entries in a map we can use the `find_opt` function:
 
-```ocaml
-# let data = data |> Stock.add Orange 30 |> Stock.add Banana 42;;
-val data : int Stock.t = <abstr>
+```ocaml=
+let has_2112_value : int option = 
+  StringMap.find_opt "leostera" lucky_numbers
+;;
 ```
 
-Now that we have some data inside our map, wouldn't it be nice
-to be able to view that data at some point? Let's begin by creating a
-simple print function.
+We can also use `find_first_opt` and `find_last_opt` if we want to use a predicate function:
 
-```ocaml
-# let string_of_fruit = function
-    | Apple -> "apple"
-    | Orange -> "orange"
-    | Banana -> "banana"
+```ocaml=
+let first_under_10_chars : (string * int) option = 
+  StringMap.find_first_opt
+    (fun key -> String.length key < 10)
+    lucky_numbers
 ;;
-val string_of_fruit : fruit -> string = <fun>
+```
+
+Note that `find_first_opt` and `find_last_opt` will return the entire entry, and not just the value.
+
+### Merging Maps
+
+To merge 2 maps you can use the `union` function, which takes 2 maps, and a function that decides how to resolve conflicting keys, and returns a new map. Note that the input maps are not modified.
 
-# let print_fruit key value =
-    print_string (string_of_fruit key ^ " " ^ string_of_int value ^ "\n")
+```ocaml=
+let all_numbers =
+    StringMap.union
+        conflict_fun
+        lucky_numbers
+        unlucky_numbers
 ;;
-val print_fruit : fruit -> int -> unit = <fun>
 ```
 
-We have here a function that will take a `fruit` key, and a quantity value,
-and print them out nicely, including a new line character at the end.
-All we need to do is to have this function applied to our mapping. Here
-is what that would look like.
-
-```ocaml
-# Stock.iter print_fruit data;;
-apple 10
-orange 30
-banana 42
-- : unit = ()
+The behavior of `StringMap.union` can be tuned based on the conflict function we pass in.
+
+For example, we can always pick the value from the first map, or choose to sum both values together (which would only work for values of type `int`), or we can even choose to ignore a value entirely if it is present in both maps.
```

Comment:
Since there are three choices, which makes for a long, slightly confusing sentence, what do you think of making these a bullet list?